### ISSUE TITLE
Host + Renter API

### ISSUE THREAD

**Issue by lukechampine** [NebulousLabs/Sia#723]
Here's my proposal:

Host
----

Queries:

* /host                 [GET]
* /host/announce        [POST]
* /host/settings        [GET]
* /host/settings        [POST]

#### /host [GET]

Function: Returns basic information about the host, including the number of
file contracts, the amount of storage remaining, and the host's profit to date.

Parameters: none

Response:
```
struct {
	storageremaining int64
	numcontracts     int
	profit           int
	potentialprofit  int
	competition      int
	unlockhash       string
	ipaddress        string
}
```

#### /host/announce [POST]

Function: Announces the host to the network as a source of storage. Generally
only needs to be called once.

Parameters: none

Response: standard

#### /host/settings [GET]

Function: Returns the current host configuration.

Parameters: none

Response:
```
struct {
	totalstorage int
	minfilesize  int
	maxfilesize  int
	minduration  int
	maxduration  int
	windowsize   int
	price        int
	collateral   int
}
```

`totalstorage` is how much storage (in bytes) the host will rent to the
network.

`minfilesize` is the minimum allowed file size.

`maxfilesize` is the maximum allowed file size.

`minduration` is the minimum amount of time a contract is allowed to last.

`maxduration` is the maximum amount of time a contract is allowed to last.

`windowsize` is the number of blocks a host has to prove they are holding the
file.

`price` is the cost (in hastings per byte) of data stored.

`collateral` is the amount of collateral the host will offer (in hastings per
byte per block) for losing files on the network.

#### /host/settings [POST]

Function: Sets the host configuration.

Parameters:
```
totalstorage int
minfilesize  int
maxfilesize  int
minduration  int
maxduration  int
windowsize   int
price        int
collateral   int
```

See the corresponding GET call for a description of each parameter.

Response: standard

----
----

The biggest change here is that `settings` is now symmetrical. Also, the old `configure` call would only modify supplied parameters; in the new scheme, all parameters are mandatory. This means that if you want to modify one field, you have to GET the old values and then POST the full updated version.

I also suggest two new endpoints, to be added some time in the future:

* /host/contracts       [GET]
* /host/contracts/$(id) [GET]

`contracts` would return a list of ids, which could be used to fetch a specific contract. (Alternatively, all contracts could be returned in one call.) There isn't much use for this right now, but I think eventually we'll want access to this information.

**Comment by DavidVorick** [comment]
I like that /host/settings is symmetric, but I think it also makes sense to return the settings field in /host [GET] instead of in /host/settings [GET] (or maybe in addition to? but the duplication doesn't seem needed)

The other thing is that it would be nice to have a host toggle.

/host/enable
/host/disable

something like that.

**Comment by lukechampine** [comment]
Right, forgot about those. I'm not sure I agree about the settings though. Symmetry is a nice property for APIs to have; if I PUT something at an endpoint, I want a subsequent GET to retrieve what I PUT.

As for the renter:

Renter
------

Queries:

* /renter/downloads     [GET]
* /renter/files         [GET]
* /renter/files/$(name) [GET]
* /renter/files/$(name) [PUT]
* /renter/files/$(name) [DELETE]
* /renter/load          [POST]
* /renter/rename        [POST]
* /renter/share         [POST]

#### /renter/downloads

Function: Lists all files that have been downloaded this session, including
downloads currently in progress.

Parameters: none

Response:
```
[]struct{
	starttime   int64
	filesize    uint64
	received    uint64
	destination string
	nickname    string
}
```
Each downloaded file is represented by the above struct.

`starttime` is the Unix time at which the download was initiated.

`filesize` is the size of the file being download.

`received` is the number of bytes downloaded thus far.

`destination` is the path that the file was downloaded to.

`nickname` is the nickname used to reference the file.

#### /renter/files [GET]

Function: Lists the status of all files.

Parameters: none

Response:
```
[]struct {
	available      bool
	uploadprogress float32
	nickname       string
	filesize       uint64
	timeremaining  types.BlockHeight (uint64)
}
```
Each uploaded file is represented by the above struct.

`available` indicates whether or not the file can be downloaded.

`uploadprogress is a percentage 0-100%, indicating how much of the file has
been uploaded. Due to redundancy, a file may become available before it is
fully uploaded.

`nickname` is the nickname used to reference the file.

`filesize` is the size of the file.

`timeremaining` indicates how many blocks the file will be available for.

#### /renter/files/$(name) [GET]

Function: Downloads a file. This call blocks until the download completes.

Parameters:
```
destination string
```
`destination` is the path that the file will be downloaded to.

Response: standard

#### /renter/files/$(name) [PUT]

Function: Uploads a file. This call blocks until the upload completes.

Parameters:
```
source string
```
`source` is the path of the file being uploaded.

Response: standard

#### /renter/files/$(name) [DELETE]

Function: Deletes a renter file entry. Does not delete any downloads or
original files, only the entry in the renter.

Parameters: none

Response: standard

#### /renter/files/load [POST]

Function: Load a '.sia' file into the renter. A '.sia' file may contain multiple files.

Parameters:
```
filename string
ascii    string
```
`filename` is the path of the '.sia' file that is being loaded.

`ascii` is the ASCII-encoded '.sia' file that is being loaded.

Either `filename` or `ascii` may be provided. If both are provided, behavior
is undefined. If neither is provided, an error is returned.

Response:
```
struct {
	filesadded []string
}
```

`filesadded` is a list of nicknames contained in the '.sia' file.

#### /renter/files/rename [POST]

Function: Rename a file. Does not rename any downloads or source files, only
renames the entry in the renter.

Parameters:
```
nickname string
newname  string
```
`nickname` is the current name of the file entry.

`newname` is the new name for the file entry.

Response: standard.

#### /renter/files/share [POST]

Function: Create a '.sia' file that can be shared with other people.

Parameters:
```
nicknames []string
filepath  string
```
`nickname` is the nickname of the file that will be shared.

`filepath` is the path of the '.sia' file that will be created. `filepath` must have the suffix '.sia'. If `filepath` is not supplied, the '.sia' file will be ASCII-encoded and returned in the response body.

Response: with `filepath`, standard. Otherwise:
```
struct {
	ascii string
}
```
`ascii` is the ASCII-encoded '.sia' file.

----
----

Controversial changes here are the use of PUT and DELETE, and conditional ASCII vs. binary .sia files. I think the case for REST semantics is much stronger in the renter, which is essentially an object store. But I'm not going to press it too hard; I'm okay with switching the PUT to POST and adding a /renter/delete endpoint.

`share` and `load` now determine whether you want ASCII or binary by looking at parameters, instead of being separate endpoints. In general I think it makes sense to use parameters to declare what encoding you want, but you may disagree.

**Comment by DavidVorick** [comment]
I would like to stay with just [GET] and [POST] if possible.

Specifying the encoding as a parameter makes sense.

I am concerned about pagination. /renter/files [GET] could return 50,000 elements, which is probably too much strain for a single api call. I'm not sure the appropriate way to handle it. In the wallet, you can request ranges of blocks. Maybe in the renter you can request an alphabetical range?

Perhaps that is thinking too far ahead, but we will want to make sure we can add pagination in a compatible way in the future.

**Comment by lukechampine** [comment]
yeah, it's a tricky question. Check out what Dropbox does: https://www.dropbox.com/developers/core/docs#metadata

One of the cool optimizations they do is to return a hash of the metadata that gets returned. Then when you make the same query, you supply the hash as well. Then the API has the ability to a "nothing changed" response code instead of the full (potentially large) response.

**Comment by DavidVorick** [comment]
We could limit the calls to returning folders, and then limit the size of a folder

**Comment by DavidVorick** [comment]
I was brainstorming the renter api today, and this is what I came up with:

/download/$(nick)
/upload -> a one-time, non-repairing, non-renewing upload
/upload/file -> watch/track/maintain a file on disk. Will do repairs, will renew
/upload/folder -> watch/track/maintain a folder on disk. Will recursively enter the folder, will follow sym links, and will not go in a loop (shouldn't be too hard to stop a loop, I think there's an AbsolutePath command)

And then of course all the renaming and stuff. Renewing and repairing in this implementation will only happen with files that are already on disk.

**Comment by lukechampine** [comment]
addressed in #891
